src/FileSystem-Core/FileReference.class.st

```diff
@@ -580,6 +575,26 @@ FileReference >> uid: uid gid: gid [
 	^filesystem file: self path uid: uid gid: gid
 ]
 
+{ #category : #versions }
+FileReference >> versionNumberFor: basename extension: extension [
+	"Answer the latest (largest) version number for the specified file.
+	0 = basename.extension exists, but nothing later.
+	nil = no file exists"
+
+	| regex maxVersion |
+
+	regex := (basename, '.(\d+).', extension) asRegex.
```

Comment:
It adds implicit dependency on regex engine. Do we know if it is correct or not? At what time regex is loaded? I suspect that it would be broken dependency.

Reply:
Another notice: can we be sure that baseline and extension parts of regex string are safe to be in regex expression? I think not. And it requires escaping which we don't have.

So I think it would be better to avoid regex with special helper method. It will also improve performance because regex in Pharo is highly unoptimized.

Reply:
Hi Dennis,

> It adds implicit dependency on regex engine. Do we know if it is correct or not? At what time regex is loaded? I suspect that it would be broken dependency.

Regex is loaded immediately after FileSystem-Memory.

I thought there were automated tests for the dependencies, and I don't remember anything being flagged.


> Another notice: can we be sure that baseline and extension parts of regex string are safe to be in regex expression? I think not. And it requires escaping which we don't have.

Can you provide an example of an unsafe file name?


> So I think it would be better to avoid regex with special helper method. It will also improve performance because regex in Pharo is highly unoptimized.

It would be nicer to optimise Regex - but since I'm not volunteering I'm not pushing the point :-)

Thanks,
Alistair

Reply:
> > It adds implicit dependency on regex engine. Do we know if it is correct or not? At what time regex is loaded? I suspect that it would be broken dependency.
> 
> Regex is loaded immediately after FileSystem-Memory.
> 
> I thought there were automated tests for the dependencies, and I don't remember anything being flagged.

I am sure we don't have tests for such kind of dependency. It is just a message send, not a class reference. 

> > Another notice: can we be sure that baseline and extension parts of regex string are safe to be in regex expression? I think not. And it requires escaping which we don't have.
> 
> Can you provide an example of an unsafe file name?

I don't know really. But #asRegex could easily fail on arbitrary string. But maybe "file name format" by itself is restricted enough to be safe for regex syntax.

Reply:
> I am sure we don't have tests for such kind of dependency. It is just a message send, not a class reference.

D'oh, of course.

> I don't know really. But #asRegex could easily fail on arbitrary string. But maybe "file name format" by itself is restricted enough to be safe for regex syntax.

It's finally clicked for me what you are talking about. :-)
Linux file names don't have any restriction (apart from / and null) so they could easily create a bad regex expression.

I'll re-write this as you suggest (although it will be after I've finished looking at ZnMimePart).

Thanks,
Alistair

Reply:
I decided to play with it and done: https://github.com/pharo-project/pharo/pull/2848